partiql-planner/src/main/kotlin/org/partiql/planner/internal/util/TypeUtils.kt

```diff
@@ -23,61 +28,70 @@ internal object TypeUtils {
         return steps.fold(type) { acc, step ->
             when (acc.code()) {
                 PType.DYNAMIC -> CompilerType(PType.dynamic())
-                PType.ROW -> acc.excludeStruct(step, lastStepOptional)
+                PType.ROW -> acc.excludeRow(step, lastStepOptional)
                 PType.STRUCT -> acc
                 PType.ARRAY, PType.BAG -> acc.excludeCollection(step, lastStepOptional)
                 else -> acc
             }
         }
     }
 
+    private fun CompilerType.PTypeField.applyExclusion(substeps: List<Rel.Op.Exclude.Step>, lastStepOptional: Boolean): CompilerType.PTypeField? {
+        val field = this
+        return if (substeps.isEmpty()) {
+            if (lastStepOptional) {
+                CompilerType.PTypeField(field.name, field.type)
+            } else {
+                null
+            }
+        } else {
+            val k = field.name
+            val v = field.type.exclude(substeps, lastStepOptional)
+            CompilerType.PTypeField(k, v)
+        }
+    }
+
     /**
-     * Applies exclusions to struct fields.
+     * Applies exclusions to `ROW` fields. Any `ROW`s that have an excluded field will also include the meta
+     * `CONTAINS_EXCLUDED_FIELD` set to true.
      *
      * @param step
      * @param lastStepOptional
      * @return
      */
-    internal fun CompilerType.excludeStruct(step: Rel.Op.Exclude.Step, lastStepOptional: Boolean = false): CompilerType {
+    internal fun CompilerType.excludeRow(step: Rel.Op.Exclude.Step, lastStepOptional: Boolean = false): CompilerType {
         val type = step.type
         val substeps = step.substeps
         val output = fields.mapNotNull { field ->
-            val newField = if (substeps.isEmpty()) {
-                if (lastStepOptional) {
-                    CompilerType.PTypeField(field.name, field.type)
-                } else {
-                    null
-                }
-            } else {
-                val k = field.name
-                val v = field.type.exclude(substeps, lastStepOptional)
-                CompilerType.PTypeField(k, v)
-            }
```

Comment:
self-review: this was moved to the above extension function, `applyExclusion`. In the prior implementation, computing the `newField` here **before** checking if the field name matches the exclude step was incorrect as we were computing the `newField` before confirming it was actually excluded.